lib/improver/tests/utilities/test_cube_units.py

```diff
@@ -34,38 +34,275 @@
+        self.assertEqual(result[0].coord('percentile').dtype, np.float32)
+        self.assertEqual(result[1].coord('air_temperature').units, 'K')
+
+    def test_subset_of_coordinates(self):
+        """Test function can enforce on a selected subset of coordinates and
+        leave all others unchanged"""
+        self.percentile_cube.coord('percentile').points = (
+            self.percentile_cube.coord('percentile').points.astype(np.int32))
+        self.percentile_cube.coord('time').convert_units(
+            'hours since 1970-01-01 00:00:00')
+        self.probability_cube.coord(
+            'air_temperature').convert_units('Fahrenheit')
+        self.probability_cube.coord('forecast_period').convert_units('h')
+
+        result = cube_units.enforce_units_and_dtypes(
+            [self.percentile_cube, self.probability_cube],
+            coords=["time", "forecast_period"])
+        self.assertEqual(result[0].coord('percentile').dtype, np.int32)
+        self.assertEqual(
+            result[0].coord('time').units, 'seconds since 1970-01-01 00:00:00')
+        self.assertEqual(
+            result[1].coord('air_temperature').units, 'Fahrenheit')
+        self.assertEqual(result[1].coord('forecast_period').units, 's')
+
+    def test_quantity_available(self):
+        """Test error raised if the named quantity is not listed in the
+        dictionary standard"""
+        self.data_cube.rename("number_of_fish")
+        self.data_cube.units = "1"
+        msg = "Name 'number_of_fish' not defined in units.py"
+        with self.assertRaisesRegex(KeyError, msg):
+            cube_units.enforce_units_and_dtypes(self.data_cube)
+
+
+# limited item dictionary to point to in smaller tests
+TEST_DICT = {
+    "time": {
+        "unit": "seconds since 1970-01-01 00:00:00",
+        "dtype": np.int64},
+    "probability": {"unit": "1"},
+    "temperature": {"unit": "K"},
+    "rainfall": {"unit": "m s-1"},
+    "rate": {"unit": "m s-1"}
+}
+
+
+class Test__find_dict_key(IrisTest):
+    """Test method to find suitable substring keys in dictionary"""
+
+    @staticmethod
+    def setUp():
+        """Redirect to dummy dictionary"""
+        cube_units.DEFAULT_UNITS = TEST_DICT
+
+    def test_match(self):
+        """Test correct identification of single substring match"""
+        result = cube_units._find_dict_key("air_temperature", "")
+        self.assertEqual(result, "temperature")
+
+    def test_probability_match(self):
+        """Test the correct substring is returned for an IMPROVER-style
+        probability cube name that matches multiple substrings"""
+        result = cube_units._find_dict_key(
+            "probability_of_air_temperature_above_threshold", "")
+        self.assertEqual(result, "probability")
+
+    def test_no_matches_error(self):
+        """Test a KeyError is raised if there is no matching substring"""
+        with self.assertRaises(KeyError):
+            cube_units._find_dict_key("snowfall", "")
```

Comment:
snowfall may well exist at some point, could this be the time for kittens to shine?

Reply:
There is no snowfall in the local test dictionary.  But maybe kittens would be clearer...